### PR TITLE
[ML] Regression dependent variable must be numeric

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
This adds a validation that the dependent variable of a regression
analysis must be numeric.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/Regression.java [hunk 1]
```diff
@@ -201,8 +202,8 @@ public boolean supportsCategoricalFields() {
     }
 
     @Override
-    public Set<String> getRequiredFields() {
-        return Collections.singleton(dependentVariable);
+    public List<RequiredField> getRequiredFields() {
+        return Collections.singletonList(new RequiredField(dependentVariable, new TreeSet<>(Types.numerical())));
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/Regression.java]: Since you are already using `Collections.singletonList` in this line, could you also use `Collections.singleton` instead of `new TreeSet<>`?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RequiredField.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+public class RequiredField {
+
+    private final String name;
+    private final SortedSet<String> types;
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RequiredField.java]: Why does it need to be sorted? Please add a comment.

**Review by przemekwitek — APPROVED** [review]
LGTM

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/Regression.java (hunk 1 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/Regression.java]: Actually, I can simply pass `Types.numerical()` directly. I'll do that.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RequiredField.java (hunk 2 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RequiredField.java]: It is to ensure types are sorted in error messages. I'll add a comment.

**Comment by dimitris-athanasiou** [comment]
@przemekwitek I have pushed a commit that addresses your feedback. However, the commit also fixes a couple other issues and it needs its own review. Would you please have a look?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java [hunk 3]
```diff
@@ -202,16 +191,64 @@ private void includeAndExcludeFields(Set<String> fields) {
...
+    private void checkFieldsHaveCompatibleTypes(Set<String> fields) {
+        for (String field : fields) {
+            Map<String, FieldCapabilities> fieldCaps = fieldCapabilitiesResponse.getField(field);
+            if (fieldCaps == null) {
+                throw ExceptionsHelper.badRequestException("no mappings could be found for field [{}]", field);
+            }
+
+            // Check all field has compatible type
+            Set<String> asSet = new HashSet<>(Collections.singleton(field));
+            removeFieldsWithIncompatibleTypes(asSet);
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java]: Personally, I would extract out the part of `removeFieldsWithIncompatibleTypes` that checks the types and re-use the extracted method here and in `removeFieldsWithIncompatibleTypes`. It would allow for more direct statement, like:
```
if (!fieldHasCompatibleTypes(field)) {
  throw ...
}
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java [hunk 4]
```diff
@@ -80,33 +69,33 @@
...
-            extractedFields = fetchFromSourceIfSupported(extractedFields);
-            if (extractedFields.getDocValueFields().size() > docValueFieldsLimit) {
-                throw ExceptionsHelper.badRequestException("[{}] fields must be retrieved from doc_values but the limit is [{}]; " +
-                    "please adjust the index level setting [{}]", extractedFields.getDocValueFields().size(), docValueFieldsLimit,
-                    IndexSettings.MAX_DOCVALUE_FIELDS_SEARCH_SETTING.getKey());
-            }
+        checkNoIgnoredFields(fields);
+        checkFieldsHaveCompatibleTypes(fields);
+        checkRequiredFields(fields);
+        checkResultsFieldIsNotPresent();
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java]: It performs the check whether or not `resultsField` is `null`. Is that ok? Do we have a test for that?

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetectorTests.java [hunk 5]
```diff
@@ -146,7 +146,55 @@ public void testDetect_GivenRegressionAndRequiredFieldMissing() {
...
+    }
+
+    public void testDetect_GivenRegressionAndRequiredFieldNotIncluded() {
+        FieldCapabilitiesResponse fieldCapabilities = new MockFieldCapsResponseBuilder()
+            .addAggregatableField("some_float", "float")
+            .addAggregatableField("some_long", "long")
+            .addAggregatableField("some_keyword", "keyword")
+            .addAggregatableField("foo", "float")
+            .build();
+        FetchSourceContext analyzedFields = new FetchSourceContext(true, new String[]  {"some_float", "some_keyword"}, new String[0]);
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetectorTests.java]: What if `foo` was in both `included` and `excluded`?

**Review by przemekwitek — APPROVED** [review]
@dimitris-athanasiou LGTM after you change although `ExtractedFieldsDetector` class is a bit hard to follow. I don't have an immediate remedy to that though :/

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java (hunk 3 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java]: Will do

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java (hunk 4 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java]: Yes, it should be ok. We just don't want to perform this validation when the resultsField is null. There is a test: `testDetect_NullResultsField`. You've added it ;-)

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetectorTests.java (hunk 5 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetectorTests.java]: `excluded` prevails. I'll add a test.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java (hunk 4 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java]: I see your point. I have removed the call to `checkResultsFieldIsNotPresent` from `detect()` because it is redundant.

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/1
